actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1085 @@
+                data_loader.append(output_batches)
+
+                # Reset
+                count = 0
+                output_batches = torch.zeros((batch_size, *obs_shape), dtype=self._dtype).to(self._device)
+            else:
+                count += 1
+                if i == num_data - 1:
+                    data_loader.append(output_batches[:count])
+
+        return data_loader
+
+    def normalize_rewards(self, rewards: list, model: PPOModel) -> list:
+        """Normalize the rewards"""
+        normalized_reward = []
+        for rew in rewards:
+            normalized_reward.append(rew / (model.reward_normalizaiton.var + 1e-8) ** 0.5)
+            self.returns = self.returns * self._cfg.discount_factor + rew
+            model.reward_normalization.update(self.returns)
+
+        return normalized_reward
+
+    def compute_gae(
+        self, rewards: torch.Tensor, values: torch.Tensor, dones: torch.Tensor, gamma: float = 0.99, lam: float = 0.95
+    ) -> torch.Tensor:
+        """Compute Generalized Advantage Estimation. See equations 11 & 12 in
+        https://arxiv.org/pdf/1707.06347.pdf
+        """
+
+        advs = []
+        with torch.no_grad():
+            gae = 0.0
+            # dones = torch.cat((dones, torch.zeros(1, 1).to(self._device)), dim=0)
+            for i in reversed(range(len(rewards))):
+                delta = rewards[i] + gamma * values[i + 1] * (1 - dones[i + 1]) - values[i]
+                gae = delta + gamma * lam * (1 - dones[i + 1]) * gae
+                advs.append(gae)
+        advs.reverse()
+        return torch.vstack(advs)
+
+
+class PPOSelfTraining(BasePPOTraining):
+    """Train PPO agent"""
+
+    async def trial_sample_sequences_producer_impl(self, sample_producer_session: SampleProducerSession):
+        """Collect sample from the trial"""
+        observation = []
+        action = []
+        reward = []
+        done = []
+        steps = []
+        player_actor_params = sample_producer_session.trial_info.parameters.actors[0]
+        player_actor_name = player_actor_params.name
+        player_environment_specs = EnvironmentSpecs.deserialize(player_actor_params.config.environment_specs)
+        player_observation_space = player_environment_specs.get_observation_space()
+        player_action_space = player_environment_specs.get_action_space()
+        obs_shape = tuple(self._cfg.image_size)[::-1]
+        async for sample in sample_producer_session.all_trial_samples():
+            previous_actor_sample = sample.actors_data[player_actor_name]
+            player_actor_name = previous_actor_sample.observation.current_player
+            actor_sample = sample.actors_data[player_actor_name]
+            obs_flat = player_observation_space.deserialize(actor_sample.observation).flat_value
+            observation_value = torch.unsqueeze(
+                torch.permute(torch.tensor(obs_flat, dtype=self._dtype).reshape(obs_shape), (2, 0, 1)), dim=0
+            )
```

Comment:
Observation space is fine. It's just  because we use the Atari wrapper where we reduce the image size down to 84x84 and we stack 4 consecutive frames together (with frame skip mechanism). In additon, for the petting zoo, we have two additional channels that represent the player index. Therefore, we have an image of 84x84x6. We need to permute this image to the input size of cnn i.e., 6x84x84